experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,694 @@
+      f::AbstractAlgebra.Generic.Frac{RET}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Given a rational map ``ϕ : X ---> Y ⊂ Spec 𝕜[y₁,…,yₙ]`` of affine schemes 
+determined by ``ϕ*(yⱼ) = fⱼ = aⱼ/bⱼ``, find the maximal open subset ``U⊂ X`` 
+to which ``ϕ`` can be extended to a regular map ``g : U → Y`` and return ``g``.
+"""
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    Y::Spec{BRT, BRET, RT, RET, MST}, 
+    f::Vector{AbstractAlgebra.Generic.Frac{RET}}
+  ) where {BRT, BRET, RT, RET, MST}
+  U, g = maximal_extension(X, f)
+  n = length(patches(U))
+  maps = Vector{SpecMor{BRT, BRET, RT, RET, MST, MST}}()
+  for i in 1:n
+    push!(maps, SpecMor(patches(U)[i], Y, [representatives(a)[i] for a in g]))
+  end
+  return SpecOpenMor(U, SpecOpen(Y), maps)
+end
+
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    Y::Spec{BRT, BRET, RT, RET, MST}, 
+    f::Vector
+  ) where {BRT, BRET, RT, RET, MST}
+  return maximal_extension(X, Y, FractionField(base_ring(OO(X))).(f))
+end
+
+function restriction(
+    f::SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2},
+    U::SpecOpen{BRT, BRET, RT, RET, MST1},
+    V::SpecOpen{BRT, BRET, RT, RET, MST2}
+  ) where {BRT, BRET, RT, RET, MST1, MST2}
+  inc = SpecOpenMor(U, domain(f), [SpecMor(W, ambient(domain(f)), gens(localized_ring(OO(W)))) for W in patches(U)])
+  help_map = compose(inc, f)
+  return SpecOpenMor(U, V, maps_on_patches(help_map))
+end
+
+identity_map(U::SpecOpen) = SpecOpenMor(U, U, [SpecMor(V, ambient(U), gens(localized_ring(OO(V)))) for V in patches(U)])
+
+function ==(f::T, g::T) where {T<:SpecOpenMor} 
+  domain(f) == domain(g) || return false
+  codomain(f) == codomain(g) || return false
+  Y = ambient(codomain(f))
+  m = length(patches(domain(f)))
+  n = length(patches(domain(g)))
+  for i in 1:m
+    for j in 1:n
+      restrict(f[i], intersect(domain(f)[i], domain(g)[i]), Y) == 
+      restrict(g[i], intersect(domain(f)[i], domain(g)[i]), Y) || return false
+    end
+  end
+  return true
+end
+
+function preimage(f::SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2},
+    Z::Spec{BRT, BRET, RT, RET, MST2}
+  ) where {BRT, BRET, RT, RET, MST1, MST2}
+  U = domain(f) 
+  X = ambient(U)
+  n = length(patches(U))
+  W = localized_ring(OO(X))
+  I = ideal(W, one(W))
+  for i in 1:n 
+    I = intersect(I, localized_modulus(OO(closure(preimage(f[i], Z), X))))
+  end
+  fZbar = subscheme(X, I)
+  return SpecOpen(fZbar, gens(U))
+end
+
+function preimage(f::SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2},
+    V::SpecOpen{BRT, BRET, RT, RET, MST2}
+  ) where {BRT, BRET, RT, RET, MST1, MST2}
+  U = domain(f)
+  X = ambient(U)
+  R = base_ring(OO(X))
+  I = ideal(R, one(R))
+  for i in 1:npatches(U)
+    I = intersect(I, saturated_ideal(ideal(OO(U[i]), pullback(f[i]).(gens(V)))))
+  end
+  return intersect(U, SpecOpen(X, gens(I)))
+end
+
+function is_non_zero_divisor(f::RET, U::SpecOpen{BRT, BRET, RT, RET}) where {BRT, BRET, RT, RET}
```

Comment:
`iszero_divisor` and `!iszero_divisor` seem more natural than `!is_non_zero_divisor` in line 670

Reply:
I don't think so. Nonzerodivisor is a frequently encountered term in CA.

Reply:
o.k.